### ISSUE TITLE
Can't pass in borrowed data to systems

### ISSUE THREAD

**Issue by eandre** [HeroesGrave/ecs-rs#24]
From the general questions thread it seems that the idea is to eventually allow systems to store references to data (very hard today since explicit lifetimes can't be used in macros, as you noted).

However, this design necessitates that systems keep a persistent reference to the data they need to access. If you want to have a RenderSystem which has access to a renderer, it likely needs mutable access and thus you either need to wrap the renderer like `Rc<RefCell<Renderer>>`, or alternatively have the RenderSystem own the renderer, but then other parts of your game cannot easily render things.

I would like to propose that one can write a custom `process()` method with a custom signature and call it yourself. However this means that you can no longer have e.g. `EntitySystem<RenderProcess>` (since you can't have a trait to satisfy with your custom `process()` signature). Instead, I think it would be good to separate the processing from the calculation of the interesting entity set. You would thus call something like (semi-pseudocode) `world.systems.render.process(world.systems.render_entities(), &mut world.data, &mut renderer)`.

Does that make sense at all?

**Comment by eandre** [comment]
I realize with this approach it will be hard to support the wrapping of systems in other systems like you have today (IntervalSystem, LazySystem) and so on, since the custom process signatures would not satisfy a trait and rust doesn't support variadic generics. Hmm.

**Comment by HeroesGrave** [comment]
Looks like `https://github.com/rust-lang/rust/pull/25509` has been delayed further. :(

Might have to look into a temporary workaround.

**Comment by Futile** [comment]
@eandre have you checked our services yet? The `&mut Datahelper<YourComponents, YourServices>` gives you mutable access to your services during `process`. Services can be anything you want, for example I keep a struct which stores textures there. Also, services are accessible from all system as well as from outside, which means that 'other parts of your game can easily render things'.

The only compromise is that the `services` field needs to own the data, so if for some reason you can't (or don't want to) hand over ownership of the renderer, this isn't a possible solution.

**Comment by HeroesGrave** [comment]
@Futile: The problem here is that services must be `'static`, and in some cases borrowed data is needed by the system. As mentioned, all the needed stuff is in the type definitions. It's just the macros can't take advantage of it yet.

**Comment by tomjakubowski** [comment]
> From the general questions thread it seems that the idea is to eventually allow systems to store references to data (very hard today since explicit lifetimes can't be used in macros, as you noted).

Not sure if this helps the specific problem here, but they can be used in macros (via the `tt` specifier and the `as_item!` hack) in some cases: https://play.rust-lang.org/?gist=2dc9313d4f78388f7dd7&version=stable

**Comment by tomjakubowski** [comment]
Also, FWIW, I don't totally know how that hack works; my understanding is that wrapping the item with `as_item!` (in the `make_impl` macro expansion)  makes the parser stop parsing the item until the expansion is totally complete, which makes it not barf on seeing a `'a`.

**Comment by HeroesGrave** [comment]
v0.23.0 adds the ability for custom processes with the `process!` macro.

Usage is as follows:
```rust
impl MySystem {
    pub fn custom_process(&mut self, custom_arg: &str, data: &mut DataHelper<_, _>) {
        do_something(custom_arg);
    }
}

// To call the function:
process!(world, my_system.custom_process(custom_arg));
```

Does this fix your problem?

**Comment by eandre** [comment]
I think so! It seems like a neat solution. Does the macro take care of passing in the `data` parameter into the `custom_process` method?

**Comment by HeroesGrave** [comment]
Yep. And if you don't need the data you should just call `world.systems.my_system.custom_process(custom_arg);`

**Comment by Ploppz** [comment]
Would it be possible to also have `entity_process!` like mentioned in https://github.com/HeroesGrave/ecs-rs/issues/13#issuecomment-231718447?
